src/components/moderation/moderation-common.ts

```diff
@@ -62,48 +62,78 @@ export const joke_responses = [
     "Didn't work. Maybe a skill issue?",
 ];
 
-// returns duration in ms
-function parse_unit(u: string) {
-    let factor = 1;
+function millis_of_time_unit(u: string) {
     switch (u) {
         case "y":
-            factor *= 365; // 365 days, fallthrough
+        case "year":
+        case "years":
+            return YEAR;
         case "d":
-            factor *= 24; // 24 hours, fallthrough
+        case "day":
+        case "days":
+            return DAY;
         case "h":
-            factor *= 60; // 60 minutes, fallthrough
+        case "hr":
+        case "hour":
+        case "hours":
+            return HOUR;
         case "m":
-            factor *= 60; // 60 seconds, fallthrough
+        case "min":
+        case "mins":
+        case "minute":
+        case "minutes":
+            return MINUTE;
         case "s":
-            factor *= 1000; // 1000 ms
-            break;
-        // Weeks and months can't be folded into the above as nicely
+        case "sec":
+        case "secs":
+        case "second":
+        case "seconds":
+            return SECOND;
         case "w":
-            factor *= 7 * parse_unit("d");
-            break;
+        case "week":
+        case "weeks":
+            return WEEK;
         case "M":
         case "mo":
-            factor *= 30 * parse_unit("d");
-            break;
+        case "month":
+        case "months":
+            return MONTH;
         default:
-            assert(false, "Unexpected unit");
+            return null;
     }
-    return factor;
 }
 
-// Duration string to milliseconds
-export function parse_duration(duration: string | null) {
-    if (duration === null) {
-        return null;
+export class ParseError extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = "ParseError";
     }
+}
+
+// Returns the corresponding duration in milliseconds,
+// or null for permanent duration.
+// Throws ParseError when parsing fails.
+export function parse_duration(duration: string) {
     const match = duration.match(duration_regex);
-    assert(match, "Failed to match against expected duration format");
-    if (duration == "perm") {
+    if (!match) {
+        throw new ParseError("Duration does not match expected pattern");
+    }
+    if (duration === "perm" || duration === "permanent") {
         return null;
-    } else {
-        const [_, n, unit] = match;
-        return parseInt(n) * parse_unit(unit);
     }
+    const [_, n, unit] = match;
+    const unit_millis = millis_of_time_unit(unit);
+    if (unit_millis == null) {
+        throw new ParseError(`Invalid time unit in duration: ${unit}`);
```

Comment:
It might be best to throw this where you `return null;` from `millis_of_time_unit`

Reply:
I did consider that, but I prefer keeping it because it makes `millis_of_time_unit` more reusable. There's no reason why it innately needs to throw, and we could use it later to e.g. verify whether a time unit is valid with `millis_of_time_unit(...) == null`.